Snippet FileManager

subdavis opened the pull request:
Snippet-based approach to compare with #125.

Includes several changes:
* Removes Breadcrumb from DataBrowser, add scoped slot for formal interface.
* Adds Snippet/MagicDataBrowser.vue (bad name, we should change it)
* Removes buttons from DataBrowser widget area, add scoped slot
* Add extra location helpers for readability.
* Remove ability for DataBrowser to have lazy location a la [v-input](https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/components/VInput/VInput.ts) -- Separated this feature into MagicDataBrowser.

I'm quite happy with how this refactor turned out, and it was a productive exercise in Vue .  Demo now consumes a batteries-included DataBrowser that looks like:

```
    girder-data-browser.pb-4(
      v-if="browserEnabled",
      :upload-multiple="multiple",
      :select-enabled="selectEnabled",
      :drag-enabled="dragEnabled",
      :new-folder-enabled="newFolderEnabled",
      :upload-enabled="uploadEnabled",
      :location.sync="location",
      @selection-changed="selected = $event")
```

This simplifies DataBrowser while also delivering value to the consumer.  Need to make sure it incorporates all necessary changes from #125

subdavis commented:
@matthewma7 Can we do this instead of #125.

matthewma7 commented:
Hi @subdavis, 
Thank you for actually create this PR to show the point.
But the MagicDataBrowser has exact same copy-paste props delegate, so the remaining difference between this one and https://github.com/girder/girder_web_components/pull/128 is the additional layer of wrapper component, the MagicDataBrowser to be specific. I am actually a little surprised that you even made the breadcrumb into a slot, does it mean the UpsertFolder needs to be changed to use a breadcrumb slot? 
This does achieve the functionality and the purpose you are proposing, but allow me to be frank and direct, I think this introduced more code to maintain and has the a not so useful additional layer of abstraction, the reason being, please allow me to repeat myself, I don't see DataBrowser is useful without a new folder, upload, breadcrumb (because the breadcrumb slot doesn't have default content, making it optional, and a data browser with breadcrumb is not that useful), or future features such as access control. Making the MagicDataBrowser the de facto DataBrowser and sole user of the DataBrowser.

zachmullen commented:
> I don't see DataBrowser is useful without a new folder, upload, breadcrumb

We already have use cases in the wild (with the old Backbone widget) that don't involve upload. We may want to present this to users to have them select an existing folder, items wouldn't even show up. I can definitely see a read-only context for using this where "new folder" / "new item" should not appear.

The breadcrumb might be universally useful, but it's quite easy to put it inside a slot, and it would accommodate use cases like selecting an item in a single preset folder.

I know we like to avoid trying to predict hypothetical use cases, but if the cost to accommodate them is so low, I don't see why we wouldn't do it.

subdavis commented:
I'm happy to have breadcrumb as the default slot if none is specified.  Formal interfaces are the goal.

matthewma7 commented:
Ok, snippet will work for my use case.
But I would like to say @zachmullen your use case I would say that use case is simply a disableBreadcrumb prop instead of leaving a whole slot open for customization. 

If we use default content, there is no tree shaking optimization. Is that the kind of optimization you mean? @zachmullen? If so, maybe no default slot needed as well. @subdavis

subdavis commented:
Stepping back for a moment:

I think my hangup is that I personally am starting to have a hard time reading and reasoning through https://github.com/girder/girder_web_components/blob/7145b9900d179c9f8ede7106f34b96ca399bccdd/src/components/DataBrowser.vue

There's some hard logic, and it's grown to a level of complexity that I'm starting to be uncomfortable with.  I'm absolutely not saying it's bad or messy code: It's just necessarily complex.  I don't think we should add glue code to join together upload, upsert, etc. in a file that is already complex. 

In vue, I like to think there are 3ish layers of complexity:

* Presentation layer stuff: just stateless rendering.
* State layer stuff which assembles the value of presentation props
* Composition layer stuff that handles interaction between those.

I'm concerned that we are unnecessarily merging complexities from state and composition layers in way that will make code comprehension hard (at least for me) in the future.  #125 would put me over the edge of feeling confident about future changes.

I think that, with this snippet approach, we could effectively put a cap on the scope of concerns in DataBrowser and develop new complexity in isolation elsewhere.

@matthewma7 @zachmullen would it be useful to have a quick call sometime today to check in and make sure all our concerns / needs are being met?

matthewma7 commented:
@subdavis The snippet approach would alleviate my pain of always copy from the App.js, so we could go with it if we all agree. 
But if you would like a quick sum up sync, I am all in for it. If @zachmullen is not available, quick sync just you and me would still be useful.

subdavis commented:
@matthewma7 this is ready for review.

* New snippet: FileManager.  I don't like the names `MagicDataBrowser` or `ConcreteDataBrowser` because I don't think they convey meaningful distinctions from `DataBrowser`, but `FileManager` draws a comparison to a desktop application where you can browse and manage your files.  Feedback needed
* `rootLocation` renamed to `rootLocationAllowed` to indicate that it's boolean and not some other kind of location-type  prop.
* reverted change to `createLocationValidator` because IMO a validator should return true/false.  Returning `root` vs `model` is confusing.
* added new helper `isRootLocation` which is used internally by `createLocationValidator` and externally.
* prefer properties named `internalProperty` to `property_` or `_property` because it's so hard to notice the underscore and recognize which thing is correct when debugging.  I've noticed `internalProperty` used in Vuetify and I think it'd be good to adopt in the future.
* Small nit fix to JobTable.vue where the progress bar is allowed to move around based on the size of other columns.  Pinned progress bar to the right-hand side of the table so it doesn't move from page to page.  Effectively a flexbox shrink.

subdavis commented:
Issues to address:

* Search doesn't navigate to user/folder 
* Upsert Breadcrumb Slot
* Should update README.md
* Selectable class name in DataTable.vue

matthewma7 commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+    location: {
+      type: Object,
+      validator: createLocationValidator(true),
+      default: () => { 'root'; },
```
Should be `() => ({ type: 'root' })`

subdavis replied:
Whoops.  Need to add a test for that too.

matthewma7 commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+      get() {
+        const { lazyLocation, girderRest } = this;
+        const { user } = girderRest;
+        if (lazyLocation) {
```
This will cause this component has some non-prop-controllable behavior that is determined by some global state. So the logic should simply be if no location prop, root by default.

subdavis replied:
Sure, if location is unspecified, default to root.

matthewma7 commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+  },
+
+  props: {
+    location: {
```
The value of location prop is not respected.

matthewma7 commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+  props: {
+    location: {
+      type: Object,
+      validator: createLocationValidator(true),
```
This component also needs the additional validator inside created()

subdavis replied:
Good catch.

matthewma7 commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+  },
+
+  asyncComputed: {
+    async defaultLocation() {
```
The default upload destination behavior is confusing. When a user is on its user root folder, then upload a file, the file won't be showing up. I think we should simply only allow uploading to the current folder.

subdavis replied:
So then should we disable the Upload button when the user is in their user root?

matthewma7 replied:
Yes, I had a computed to check if the current location is a folder. So, even if upload is enabled, but the current location is not a folder, the button won't show.

matthewma7 commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+          })).data[0];
+        } catch ({ response }) {
+          if (response) {
+            this.error = response.data.message;
```
Seems like a copy-paste error since there is no place for the error to show up.

subdavis commented on src/components/Presentation/DataTable.vue:
```diff
@@ -86,7 +86,7 @@ v-data-table.girder-data-table(
       td.pl-3(colspan="2")
         span.text-container.secondary--text.text--darken-3.nobreak(
-            :class="{ selectable: selection && props.item._modelType !== 'item' }",
+            :class="{ selectable: props.item._modelType !== 'item' }",
```
This is a class that adds the pointer hover style.  It should be present regardless of whether check-boxes are there or not, right?

matthewma7 replied:
Oh, ok, that's an intentional change. I must have mixed it with the other selectable we have.

subdavis commented on src/components/DataBrowser.vue:
```diff
@@ -294,7 +285,7 @@ girder-data-table.girder-file-browser(
     :total-items="totalItems",
     :loading="loading",
-    :selection="selection_",
+    :selection="internalSelectable",
```
> selection -> selectable because I think it was overlooked in #125

> prefer properties named internalProperty to property_ or _property because it's so hard to notice the underscore and recognize which thing is correct when debugging. I've noticed internalProperty used in Vuetify and I think it'd be good to adopt in the future.

I missed the rename on the left hand side, but the right hand side was intentional.

matthewma7 replied:
Oh, yes. The right side is good. I mean the left side.

subdavis replied:
I misread your first comment.  ty.

subdavis commented on src/components/Breadcrumb.vue:
```diff
@@ -62,7 +63,7 @@ export default {
           breadcrumb.unshift(this.extractCrumbData(data));
         }
-        if (this.rootLocation) {
+        if (rootLocationAllowed) {
```
I think this is a matter of personal preference.  Inside computed properties, I like to put all "dependencies" at the top as documentation.  I find myself scanning computed functions for dependencies much more often than resolving local variables, particularly if I know it's already a prop, so that's the case this optimizes for.

That being said, I don't care that strongly.  Can revert if you'd like.

matthewma7 replied:
It would be great if you would, be because a simple `this.rootLocationAllowed` is enough here.

subdavis replied:
After changing this, I noticed that the reactivity system was no longer able to depend on rootLocationAllowed.  Checking and unchecking in the demo didn't work anymore.  Reverting fixed it.

I'll probably keep putting dependencies at the top like this in the future.  I don't know why AsyncComputed works differently for this case.

matthewma7 replied:
Oh, make sense it doesn't work. All dependency of async computed needs to be accessed before the first async function. My request accidentally exactly broke it. 😂

subdavis commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+<script>
```
> The access control will be using this as well

What do you mean by that?  Nothing within the library should be importing FileManager.

Can move this into components if you want, I just though it would be good to separate so other components don't try to import from snippet/.

matthewma7 replied:
I mean, the FileManager is not a one-off component, more feature will be added to it. Such as the Access Control in the future PR, so it is a core component that deserves to live in the components folder?

subdavis replied:
Sure, so FileManager may import other things, but other things will never import it.  It's a one way relationship, which is why I separated it, just like `views/` and `components/` are separated in regular projects.

It's easy to remember not to import FileManager when it's just one thing, but if there were 6, it's nice to have the directory separation as a red flag during code review.

matthewma7 replied:
Oh, I see what you mean. You meant that anything under component is `reusable` component, this snippet is on the contrary, final. 
Make sense, let's keep it in the snippet.

subdavis commented on src/components/Snippet/FileManager.vue:
```diff
@@ -0,0 +1,215 @@
+      validator: createLocationValidator(true),
+      default: () => { 'root'; },
+    },
+    rootLocationAllowed: {
```
> it could have an invalid initial state. 

Not sure I understand.   The created hook check (that I need to add) will throw if location is root and this prop is false.  What is the consequence of making this default to false?

If it defaults to true and explicitly set it as false but pass in a root location, don't you end up in the same invalid state?

matthewma7 replied:
I meant that because rootLocationAllowed is an optional prop if the consumer doesn't set it and don't set the location, the location will default to root, and the allowed will be false, making the FileManager state invalid at the start. A component should not be invalid if an optional prop is not set. 

> If it defaults to true and explicitly set it as false but pass in a root location, don't you end up in the same invalid state?
Good question, in that case, it is the consumer who disabled the root and passes a root location (or don't pass since it's the default), so we could use one of the validators to throw an error, and the consumer should take the blame because it's their own action.

subdavis replied:
Ok, I agree with that.  Will change `rootLocationAllowed` to `rootLocationDisabled` universally.

matthewma7 replied:
Not necessary universal, because `DataBrowser` and `Breadcrumb` won't have this `initial invalid state` issue, change them universally if you would like, otherwise, just change this one is ok to me.

subdavis commented:
@matthewma7 thanks for the detailed review!  Probably won't get around to making your requested changes until tomorrow.   I'll poke you when I'm done.

subdavis commented on src/utils/locationHelpers.js:
```diff
@@ -0,0 +1,43 @@
+  } else if (['user', 'collection', 'folder'].indexOf(locationType) >= 0) {
+    return false;
+  }
+  throw new Error(`Unrecognized location: ${locationType}`);
```
I think we do.  isRootLocation implies the argument is a location.  If the argument is not a location, the developer made mistake.

subdavis replied:
I'm actually wrong, this should be true/false for use in templates.

subdavis commented:
Changes pushed based on our discussion.  

Changed the hover pointer vs regular cursor behavior markedly: it has never been correct since I wrote it the first time, and it was bothering me a lot.  New rules are sorta hard to describe, but the good news is you don't end up getting the text cursor hover anymore.

matthewma7 commented:
Hi @subdavis Thanks for the updates. 
I have one last feedback. Currently, the FileManager support no location props or location.sync props, but doesn't support :location without .sync. 
Without .sync modifier, the location will be locked. I think the .sync should be optional, not required. 

This is a little hard to describe, so I created a [demo branch](https://github.com/girder/girder_web_components/commit/5a1580e733d95afdf4001c90c04ce2fa994fa032), please give it a try before and after reverting the changes on FileManager.

subdavis commented:
> Without .sync modifier, the location will be locked. I think the .sync should be optional, not required.

This is intended behavior, a feature of the browser.   It should be possible for a consumer to lock the location by not using sync, and to force the data browser into whatever location it wants.  The demo branch you created removes that ability.

matthewma7 commented:
>This is intended behavior, a feature of the browser. The demo branch you created removes that ability.
A non-interactive data browser should be supported.

I didn't aware that behavior is on purpose. I was thinking that Vue is not real two-way binding, and in Vuetify when `:value` is used instead of `v-model`, it doesn't lock the state, but only set the initial state, then there is another prop to control disable states, such as Tooltip or Dialog.  
If you prefer this behavior, I have no strong opinion.

subdavis commented:
> in Vuetify when :value is used instead of v-model

Yeah I see what you mean, thanks for those examples.  I think this is a rare case where I actually disagree with the choice made by Vuetify.  `v-dialog` works the way you're describing, and I actually consider it a bug.  If I set `:value="true"` on my dialog and it _still_ doesn't stay open when I click outside the box, that's frustrating and violates the principle of least astonishment.

There are components that work the other way, like [v-stepper](https://vuetifyjs.com/en/components/steppers).  If you use `:value` instead of `v-model`, you get locked into the current step.

Thanks for the review Matthew.  I realize I forgot to request a review from @zachmullen so I'll give him a bit to look over this before merging.

zachmullen commented:
I can review this next week.

subdavis commented:
@zachmullen bump :)  

Let me know if I can help in any way.

zachmullen commented on src/components/Presentation/DataTable.vue:
```diff
@@ -111,6 +127,10 @@ v-data-table.girder-data-table(
 
+  // .default-cursor:hover {
+  //   cursor: default;
+  // }
```
Remove dead code

subdavis commented:
Its been a few weeks and I wanted to take one last look at this.  I wasn't happy with the demo layout, so I tweaked it.  I would have just rolled this into another PR, but I had already pretty much rewritten demo/App.vue before the last round of changes.

## Logged in view

(With jobs enabled -- jobs is disabled by default)
![Screenshot_20190812_141205](https://user-images.githubusercontent.com/4214172/62887590-8c70bf80-bd0b-11e9-904d-7d624e4bd241.png)

## Logged out view

![Screenshot_20190812_141052](https://user-images.githubusercontent.com/4214172/62887591-8c70bf80-bd0b-11e9-9092-2d84f6c5be00.png)

The reason for these changes is 

1) The data browser is now usable by logged out users, so I'd like to expose this component to anyone who doesn't have a DKC account and would otherwise exit the page immediately. I also think seeing the collections will give kitware folks more confidence that they're logging into DKC.
2) I don't personally like Vuetify's dialogs anymore, and would prefer to use them sparingly for cases where user input is required to complete an action.  They're clunky and generally make poor use of space on the page.  (This is more related to replacing the job list widget dialog than the login dialog)
